src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -15,547 +15,442 @@
-        {
-            head = node;
-            tail = node;
-        }
+        addLast(node);
         unreferenced++;
     }
 
-    private <K, V> void updateSize(Node<K, V> node, ToLongFunction<V> estimator)
-    {
-        bytesCached += node.estimatedSizeOnHeapDelta(estimator);
-    }
-
-    // don't evict if there's an outstanding save result. If an item is evicted then reloaded
-    // before it's mutation is applied, out of date info will be loaded
-    private boolean canEvict(Node<?, ?> node)
+    @SuppressWarnings("unchecked")
+    private <K, V> void maybeUpdateSize(AccordCachingState<?, ?> node, ToLongFunction<?> estimator)
     {
-        Invariants.checkState(node.references == 0);
-        return node.state() == FAILED || !hasActiveAsyncResult(saveResults, node.key());
+        if (node.shouldUpdateSize())
+            bytesCached += ((AccordCachingState<K, V>) node).estimatedSizeOnHeapDelta((ToLongFunction<V>) estimator);
     }
 
-    private void maybeEvict()
+    /*
+     * Roughly respects LRU semantics when evicting. Might consider prioritising keeping MODIFIED nodes around
+     * for longer to maximise the chances of hitting system tables fewer times (or not at all).
+     */
+    private void maybeEvictSomeNodes()
     {
         if (bytesCached <= maxSizeInBytes)
             return;
 
-        Node<?, ?> current = tail;
-        while (current != null && bytesCached > maxSizeInBytes)
+        Iterator<AccordCachingState<?, ?>> iter = this.iterator();
+        while (iter.hasNext() && bytesCached > maxSizeInBytes)
         {
-            Node<?, ?> evict = current;
-            current = current.prev;
-
-            // TODO (expected, efficiency): can this be reworked so we're not skipping unevictable nodes everytime we try to evict?
-            if (!canEvict(evict))
-                continue;
+            AccordCachingState<?, ?> node = iter.next();
+            checkState(node.references == 0);
 
-            evict(evict, true);
+            /*
+             * TODO (expected, efficiency):
+             *    can this be reworked so we're not skipping unevictable nodes everytime we try to evict?
+             */
+            Status status = node.status(); // status() call completes (if completeable)
+            switch (status)
+            {
+                default: throw new IllegalStateException("Unhandled status " + status);
+                case LOADED:
+                    unlink(node);
+                    evict(node);
+                    break;
+                case MODIFIED:
+                    // schedule a save to disk, keep linked and in the cache map
+                    Instance<?, ?, ?> instance = instanceForNode(node);
+                    node.save(saveExecutor, instance.saveFunction);
+                    maybeUpdateSize(node, instance.heapEstimator);
+                    break;
+                case SAVING:
+                    // skip over until completes to LOADED or FAILED_TO_SAVE
```

Comment:
We should preserve the TODO about skipping unselectable nodes

Reply:
I think it's sufficiently preserved, above, just above the switch statement? Unless you mean something else.

Reply:
heh, oh right, missed it there :)